### PR TITLE
refactor(core): split monolithic files into focused modules

### PR DESCRIPTION
**Pull request by cantalupo555**
## Summary
Refactors the monolithic core files into focused, single-responsibility modules. The original `fsrs.go` (222 lines) and `parameters.go` (large) have been split into 11 smaller, well-organized files.

## Motivation
The core module had grown unwieldy with mixing concerns: error definitions, rescheduling logic, rollback handling, fuzz generation, arithmetic utilities, and scheduling steps were all tangled together. This refactoring improves code organization, maintainability, and makes each file's purpose immediately clear.

## Changes Made
- Extracted error types (`ErrManualStateRequired`, etc.) into dedicated `errors.go`
- Extracted `Forget` method into dedicated `forget.go`
- Extracted `Rollback` method into dedicated `rollback.go`
- Extracted `Reschedule` family of methods into dedicated `reschedule.go`
- Extracted fuzz-related logic from `parameters.go` into dedicated `fuzz.go`
- Extracted arithmetic utilities from `parameters.go` into dedicated `arithmetic.go`
- Extracted step/delay scheduling utilities from `scheduler_basic.go` into dedicated `steps.go`
- Simplified `fsrs.go` to core struct and thin wrappers (`Repeat`, `Next`, `GetRetrievability`)
- Simplified `parameters.go` to core parameter handling only
- Expanded test coverage in `fsrs_test.go` (Reschedule edge cases from 24 to 31 subtests)

## Impact
- No functional changes — all public APIs and behavior remain identical
- Performance: same (pure refactoring, no algorithmic changes)
- Code quality: improved — each module now has a single, clear responsibility
- Breaking changes: **None** — backward-compatible reorganization

## Testing
- [x] All existing tests pass
- [x] No behavioral changes

### CODE REVIEW

**Comment by ishiko732** [comment]
I might need a bit of time to review, but it won’t take too long.

**Comment by cantalupo555** [comment]
@ishiko732 
The refactoring related to issue #53 will be addressed in future pull requests.
It hasn't been implemented in this PR yet, since those changes were already well underway before the previous review.

**Comment by cantalupo555** [comment]
> @ishiko732 The refactoring related to issue #53 will be addressed in future pull requests. It hasn't been implemented in this PR yet, since those changes were already well underway before the previous review.


All done, I've sent it to https://github.com/open-spaced-repetition/go-fsrs/pull/63

I think that once all the PRs have been merged, we can release a release candidate for version v4.🫡

There's still a lot of work ahead, but as far as the scheduler is concerned, I think we're well on our way.

File: rollback.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+		return Card{}, ErrManualRating
+	}
+	if log.Rating < Again || log.Rating > Easy {
+		return Card{}, ErrInvalidRating
+	}
+	result := card
+	result.State = log.State
+	result.Stability = log.Stability
+	result.Difficulty = log.Difficulty
+	result.ElapsedDays = log.ElapsedDays
```
Comment (ishiko732) [inline rollback.go]: I’m planning to remove ElapsedDays in the next major version of ts-fsrs, since it changes over time and probably shouldn’t be stored persistently.

You may want to consider removing ElapsedDays from both card and log as part of this v4 upgrade.

https://github.com/open-spaced-repetition/ts-fsrs/blob/d3cb3c69aa6109e11245eb85ec0f8ba7ead407a4/packages/fsrs/src/models.ts#L54-L57

File: arithmetic.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func (p *Parameters) initDifficulty(r Rating) float64 {
+	return constrainDifficulty(p.W[4] - math.Exp(p.W[5]*float64(r-1)) + 1)
+}
+
+func (p *Parameters) initDifficultyRaw(r Rating) float64 {
+	return p.W[4] - math.Exp(p.W[5]*float64(r-1)) + 1
+}
+
```
Comment (ishiko732) [inline arithmetic.go:55]:
```go
func (p *Parameters) initDifficulty(r Rating, clamp bool) float64 {
	d0 = ...
	if (!clamp){
		return constrainDifficulty(d0)
	}
	return d0
}

```

initDifficulty can be simplified.

- https://github.com/open-spaced-repetition/py-fsrs/pull/133
- https://github.com/open-spaced-repetition/py-fsrs/pull/133#discussion_r2348337661

File: arithmetic.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	return constrainStability(s * sinc)
+}
+
+func (p *Parameters) meanReversion(init float64, current float64) float64 {
+	return p.W[7]*init + (1-p.W[7])*current
+}
+
+func (p *Parameters) nextRecallStability(d float64, s float64, r float64, rating Rating) float64 {
+	d = constrainDifficulty(d)
+	s = constrainStability(s)
```
Comment (ishiko732) [inline arithmetic.go]: This seems a bit redundant. card should already ensure the output values are within the valid range, so there’s no need to clamp them again here.

File: arithmetic.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		(math.Exp((1-r)*p.W[10])-1)*
+		hardPenalty*
+		easyBonus)
+
+	return constrainStability(newS)
+}
+
+func (p *Parameters) nextForgetStability(d float64, s float64, r float64) float64 {
+	d = constrainDifficulty(d)
+	s = constrainStability(s)
```
Comment (ishiko732) [inline arithmetic.go]: Same issue here

**Review by ishiko732 — APPROVED** [review]
LGTM!

File: rollback.go (hunk 1 above)
Comment (cantalupo555) [reply rollback.go]: Thanks for letting me know; it really is redundant.
